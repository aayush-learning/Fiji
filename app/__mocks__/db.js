const usersDB = {
  sync: () => {
    return {rows: []}
  },
  allDocs: () => {
    return {rows: []}
  }
}

export { usersDB }