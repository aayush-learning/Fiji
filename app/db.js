import PouchDB from 'pouchdb-react-native';

// these settings let us use cuids
// see: https://github.com/ericelliott/cuid/issues/54
global.navigator.mimeTypes = '';
global.navigator.userAgent = 'reactnative';    

const fijiDB = new PouchDB('fiji', { adapter: 'asyncstorage' });
console.log(fijiDB.adapter)
fijiDB.info().then(function (info) {
  console.log(info);
})
export default fijiDB