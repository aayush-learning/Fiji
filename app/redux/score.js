import { contentDB, remoteContentDB } from '../db'

const FETCH_GAME_HIGH_SCORES_REQUEST = 'Fiji/game/FETCH_GAME_HIGH_SCORES_REQUEST'
const FETCH_GAME_HIGH_SCORES_SUCCESS = 'Fiji/game/FETCH_GAME_HIGH_SCORES_SUCCESS'
const FETCH_GAME_HIGH_SCORES_FAILURE = 'Fiji/game/FETCH_GAME_HIGH_SCORES_FAILURE'
const ADD_SCORE = 'Fiji/game/ADD_SCORE'

export const initialState = {
  isFetching: false,
  gameHighScores: [],
  myScore: 0,
}

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_GAME_HIGH_SCORES_REQUEST:
      return {
        ...state,
        isFetching: true
      }
    case FETCH_GAME_HIGH_SCORES_SUCCESS:
      return {
        ...state,
        isFetching: false,
        gameHighScores: action.gameHighScores
      }
    case FETCH_GAME_HIGH_SCORES_FAILURE:
      return {
        ...state,
        isFetching: false,
        gameHighScores: []
      }
    case ADD_SCORE:
      return {
        ...state,
        myScore: state.myScore + action.myScore
      }
    default:
      return state
  }
}

export const fetchGameHighScoresRequest = () => ({
  type: FETCH_GAME_HIGH_SCORES_REQUEST
})

export const fetchGameHighScoresSuccess = (gameHighScores) => ({
  type: FETCH_GAME_HIGH_SCORES_SUCCESS,
  gameHighScores
})

export const fetchGameHighScoresFailure = () => ({
  type: FETCH_GAME_HIGH_SCORES_FAILURE
})

export const addMyScore = (myScore) => ({
  type: ADD_SCORE,
  myScore
})

export const fetchGameHighScores = (game_id) => async (dispatch, getState) => {
  try {
    dispatch(fetchGameHighScoresRequest())
    const result = await contentDB.allDocs({
      startkey: 'high-score:' + game_id,
      endkey: 'high-score:' + game_id + '\ufff0',
      include_docs: true
    })
    const highScores = result.rows
      .map(function (row) { return row.doc })
      .sort((a, b) => {
      return b.score - a.score
    })
    dispatch(fetchGameHighScoresSuccess(highScores))
  } catch (error) {
    console.log('fetchGameHighScores: ' + error)
    dispatch(fetchGameHighScoresFailure())
  }
}

export const finalizeScore = (user_id, game_id, myScore) => async (dispatch, getState) => {
  try {
    const currentHighScore = await contentDB.get('high-score:' + game_id + ':' + user_id)
    if (myScore > currentHighScore) {
      const result = await contentDB.put({
        _id: 'high-score:' + game_id + ':' + user_id,
        user_id,
        myScore
      })
      console.log(result)
    }
  } catch (error) {
    console.log('finalizeScore: ' + error)
  }
}

export default reducer