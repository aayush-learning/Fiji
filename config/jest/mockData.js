import { fetchUsersFailure } from '../../app/redux/users';

const contentJson = require('../seed/content.json')
const usersJson = require('../seed/users.json')
const userAliceJson = require('../seed/user_alice.json')
const storyJson = require('../seed/story.json')

export const titles = contentJson['docs'].filter(function (doc) {
  return doc._id.startsWith('storytitle:')
})

export const stories = storyJson['docs'].filter(function (doc) {
  return doc._id.startsWith('story:')
})

export const games = contentJson['docs'].filter(function (doc) {
  return doc._id.startsWith('game:')
})

export const gameThemeData = contentJson['docs'].filter(function (doc) {
  return doc._id.startsWith('gametheme:')
})

export const stickersData = contentJson['docs'].filter(function (doc) {
  return doc._id.startsWith('sticker:caterpillar:')
})

export const stickerPacksData = contentJson['docs'].filter(function (doc) {
  return doc._id.startsWith('sticker-pack:')
})

export const highScoresData = contentJson['docs'].filter(function (doc) {
  return doc._id.startsWith('high-score:')
})

export const story_alice_in_wonderland = storyJson['docs'].filter(function (doc) {
  return doc._id == 'story:alice-in-wonderland'
})[0]

export const users = usersJson['docs']

export const messages = userAliceJson['docs'].filter(function(doc) {
  return doc._id.startsWith('chat:mad-hatter:')
})