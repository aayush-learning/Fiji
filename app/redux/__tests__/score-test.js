jest.mock('../../db')

import { contentDB } from '../../db'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import score, {
  initialState,
  fetchGameHighScoresRequest,
  fetchGameHighScoresSuccess,
  fetchGameHighScoresFailure,
  addScore,
  fetchGameHighScores
} from '../score'
import { highScoresData } from '../../../config/jest/mockData'

// test reducers

it('returns the same state on an unhandled action', () => {
  expect(score(initialState, { type: '_NULL' })).toMatchSnapshot()
})

it('handles FETCH_GAME_HIGH_SCORES_REQUEST action', () => {
  expect(score(initialState, fetchGameHighScoresRequest())).toMatchSnapshot()
})

it('handles FETCH_GAME_HIGH_SCORES_SUCCESS action', () => {
  expect(score(initialState, fetchGameHighScoresSuccess(highScoresData))).toMatchSnapshot()
})

it('handles FETCH_GAME_HIGH_SCORES_FAILURE action', () => {
  expect(score(initialState, fetchGameHighScoresFailure())).toMatchSnapshot()
})

// test actions

it('creates a FETCH_GAME_HIGH_SCORES_REQUEST action', () => {
  expect(fetchGameHighScoresRequest()).toMatchSnapshot()
})

it('creates a FETCH_GAME_HIGH_SCORES_SUCCESS action', () => {
  expect(fetchGameHighScoresSuccess(highScoresData)).toMatchSnapshot()
})

it('creates a FETCH_GAME_HIGH_SCORES_FAILURE action', () => {
  expect(fetchGameHighScoresFailure()).toMatchSnapshot()
})

// test thunks

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

it('handles fetchGameHighScores action', async () => {
  const store = mockStore(initialState)
  contentDB.allDocs = jest.fn()
    .mockReturnValue({
      rows: highScoresData.map(function(doc) {
        return { doc }
      })
    })
  await store.dispatch(fetchGameHighScores('game:memory-matching'))
  expect(store.getActions()).toMatchSnapshot()
})