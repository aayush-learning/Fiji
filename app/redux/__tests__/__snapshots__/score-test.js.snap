// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a FETCH_GAME_HIGH_SCORES_FAILURE action 1`] = `
Object {
  "type": "Fiji/game/FETCH_GAME_HIGH_SCORES_FAILURE",
}
`;

exports[`creates a FETCH_GAME_HIGH_SCORES_REQUEST action 1`] = `
Object {
  "type": "Fiji/game/FETCH_GAME_HIGH_SCORES_REQUEST",
}
`;

exports[`creates a FETCH_GAME_HIGH_SCORES_SUCCESS action 1`] = `
Object {
  "gameHighScores": Array [
    Object {
      "_id": "high-score:game:memory-matching:mad-hatter",
      "score": 20,
      "user_id": "mad-hatter",
    },
    Object {
      "_id": "high-score:game:memory-matching:white-rabbit",
      "score": 30,
      "user_id": "white-rabbit",
    },
    Object {
      "_id": "high-score:game:memory-matching:alice",
      "score": 10,
      "user_id": "alice",
    },
  ],
  "type": "Fiji/game/FETCH_GAME_HIGH_SCORES_SUCCESS",
}
`;

exports[`handles FETCH_GAME_HIGH_SCORES_FAILURE action 1`] = `
Object {
  "gameHighScores": Array [],
  "isFetching": false,
  "score": 0,
}
`;

exports[`handles FETCH_GAME_HIGH_SCORES_REQUEST action 1`] = `
Object {
  "gameHighScores": Array [],
  "isFetching": true,
  "score": 0,
}
`;

exports[`handles FETCH_GAME_HIGH_SCORES_SUCCESS action 1`] = `
Object {
  "gameHighScores": Array [
    Object {
      "_id": "high-score:game:memory-matching:mad-hatter",
      "score": 20,
      "user_id": "mad-hatter",
    },
    Object {
      "_id": "high-score:game:memory-matching:white-rabbit",
      "score": 30,
      "user_id": "white-rabbit",
    },
    Object {
      "_id": "high-score:game:memory-matching:alice",
      "score": 10,
      "user_id": "alice",
    },
  ],
  "isFetching": false,
  "score": 0,
}
`;

exports[`handles fetchGameHighScores action 1`] = `
Array [
  Object {
    "type": "Fiji/game/FETCH_GAME_HIGH_SCORES_REQUEST",
  },
  Object {
    "gameHighScores": Array [
      Object {
        "_id": "high-score:game:memory-matching:white-rabbit",
        "score": 30,
        "user_id": "white-rabbit",
      },
      Object {
        "_id": "high-score:game:memory-matching:mad-hatter",
        "score": 20,
        "user_id": "mad-hatter",
      },
      Object {
        "_id": "high-score:game:memory-matching:alice",
        "score": 10,
        "user_id": "alice",
      },
    ],
    "type": "Fiji/game/FETCH_GAME_HIGH_SCORES_SUCCESS",
  },
]
`;

exports[`returns the same state on an unhandled action 1`] = `
Object {
  "gameHighScores": Array [],
  "isFetching": false,
  "score": 0,
}
`;
